File: api/src/main/java/com/hearlers/gateway/presentation/http/v1/auth/AuthDtoMapper.java
package com.hearlers.gateway.presentation.http.v1.auth;

import com.hearlers.gateway.auth.model.AuthInfo;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.NullValueCheckStrategy;
import org.mapstruct.ReportingPolicy;

import com.hearlers.api.proto.v1.message.User;
import com.hearlers.api.proto.v1.model.AuthChannel;
import com.hearlers.api.proto.v1.model.AuthUser;
import com.hearlers.api.proto.v1.model.UserProfile;
import com.hearlers.api.proto.v1.service.InitializeUserResponse;
import com.hearlers.api.proto.v1.service.SaveRefreshTokenRequest;

@Mapper(
        unmappedTargetPolicy = ReportingPolicy.IGNORE,
        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS
)
public interface AuthDtoMapper {

    // User -> UserResponseDto
    AuthDto.UserResponseDto of(User user);

    // UserProfile -> UserProfileResponseDto
    AuthDto.UserProfileResponseDto of(UserProfile userProfile);

    // AuthUser -> SaveRefreshTokenRequest
    default SaveRefreshTokenRequest ofAuthUserAndToken(AuthUser authUser, AuthInfo.TokenInfo token) {
        return SaveRefreshTokenRequest.newBuilder()
                .setUserId(authUser.getUserId())
                .setToken(token.getRefreshToken())
                .setExpiresAt(token.getRefreshTokenExpiresAt().toString())
                .build();
    }

    // TokenInfo -> TokenResponseDto
    AuthDto.TokenResponseDto of(AuthInfo.TokenInfo token);
}

File: api/src/main/kotlin/com/hearlers/gateway/presentation/http/v1/auth/mapper/AuthDtoMapper.kt
package com.hearlers.gateway.presentation.http.v1.auth.mapper

import com.hearlers.api.proto.v1.model.AuthUser
import com.hearlers.api.proto.v1.model.User
import com.hearlers.api.proto.v1.model.UserProfile
import com.hearlers.api.proto.v1.service.SaveRefreshTokenRequest
import com.hearlers.gateway.auth.model.AuthInfo
import com.hearlers.gateway.presentation.http.v1.auth.dto.AuthDto

object AuthDtoMapper {

    fun of(user: User): AuthDto.UserResponse {
        return AuthDto.UserResponse(
            user.id,
            user.nickname,
            of(user.userProfile),
            user.createdAt,
            user.updatedAt,
            user.deletedAt
        )
    }

    fun of(userProfile: UserProfile): AuthDto.UserProfileResponse {
        return AuthDto.UserProfileResponse(
            profileImage = userProfile.profileImage,
            phoneNumber = userProfile.phoneNumber,
            gender = userProfile.gender,
            birthday = userProfile.birthday,
            introduction = userProfile.introduction,
            mbti = userProfile.mbti,
            createdAt = userProfile.createdAt,
            updatedAt = userProfile.updatedAt,
            deletedAt = if (userProfile.hasDeletedAt()) userProfile.deletedAt else null
        )
    }

    fun ofAuthUserAndToken(authUser: AuthUser, token: AuthInfo.TokenInfo): SaveRefreshTokenRequest {
        return SaveRefreshTokenRequest.newBuilder()
            .setUserId(authUser.userId)
            .setToken(token.refreshToken)
            .setExpiresAt(token.refreshTokenExpiresAt.toString())
            .build()
    }

    fun of(token: AuthInfo.TokenInfo): AuthDto.TokenResponse {
        return AuthDto.TokenResponse(
            accessToken = token.accessToken,
            refreshToken = token.refreshToken,
            accessTokenExpiresAt = token.accessTokenExpiresAt.toString(),
            refreshTokenExpiresAt = token.refreshTokenExpiresAt.toString()
        )
    }
}